ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+# 40. Workspace Abstraction
+
+Date: 2024-9-26
+
+## Status
+
+WiP
+
+## Context
+
+Workspaces are being handled differently today in our community and in our downstream deployments. 
+
+### Downstream
+
+We have a workspace CRD defined with a controller that interfaces with KubeSaw as the NSaaS implementation.
+
+Kubesaw provides multiple features but mainly namespace provisioning, multi-cluster placement algorithms, syncronization of tiers (Quotas, Roles, RoleBindings and ConfigMaps) and user lifecycle management.
+
+Workspace CRD serves as an abstraction to be able to make changes to a workspace or create new workspaces for a specific users. 
+
+This abstraction also enable Konflux to implement other features in the future like for example limmiting the amount of workspaces per user, allowing workspace delete self-service and adding workspace delete protection.
+
+
+### Upstream
+
+There is not support for multi-cluster for a single instance of Konflux. The Idea here is to require a new deployment of konflux of there is a need to scale beyond a single cluster. Therefore, there is a one to one relationship between konflux instance and k8s cluster.
+
+With this contraint in place workspace concept can be implemented differently
+We relay on the ability to create namespaces directly. Which requires access to all authenticated users to have premissions to create a namespace.
+
+The boostrapping of the namespace created is delegate to a GitOps workflow that will automatically create an application with the necesary resources, including RoleBinding and Quotas and will continue to synch this through the lifecycle of the project.
+
+To be able to see exisiting workspaces for a specific user, we developed a small abstraction called workspace manager, which is able to determine what namepsace have a rolebinding providing access to the current user.
+
+There is currently no way to delete a workspace self-service. User will have to make a request to a konflux admin to delete a specific workspace
+
+### Namespace Limitations
+
+Kubernetes namespaces where not designed for self-service provisioning in a multi-tenant.
+
+Access for namespace list and create operation is all namepaces or none. This affects multi-tenancy if self-service is desired.
+
+Access to get and delete can be granted via RBAC once the namespace is created. This will required admin access which also affect self-service.
+
+Allowing access to specific Roles or ClusterRoles via RoleBinding can be done by an Admin, also braking self-service. This is required to be able to execute Konflux actions inside the namespace.
+
+## Decision
```

Comment:
> 1. Can you please provide details about how this will get implemented? It would be nice to understand how it will work for each operation.

Below you will find more details for the design proposal be since you asked I will answer here and we can create a list once everyone ask their questions.
> 2. The `workspace` resource is name spaced or is it a cluster resource?

Namepsace scoped

> 3. What is the schema for the workspace resource?

https://github.com/konflux-workspaces/workspaces/blob/main/doc/book/src/rest-api/crds.md
We though it was not needed to get that deep but if you want to explore that

> 4. Do we need to provide a limit to the amount of namespace a user can create? from where this requirement is coming from? why in your opinion it's an issue that Konflux should solve?

I think it needs to be limited, in any self-service multi-tenant services there are sensible limits. Maybe we don't want to solve but need to provide the ability to solve as admins will likely don't want their Clusters to be either attacked or abused.
  
> 5. Is it introducing just a controller or also a REST service?

Both

Reply:
| https://github.com/konflux-workspaces/workspaces/blob/main/doc/book/src/rest-api/crds.md
We though it was not needed to get that deep but if you want to explore that

how each field in the status section is calculated?

Reply:
> | https://github.com/konflux-workspaces/workspaces/blob/main/doc/book/src/rest-api/crds.md We though it was not needed to get that deep but if you want to explore that
> 
> 1. how each field in the status section is calculated?

Two fields right now owner and space. Owner is the user making the request to create workspace, space is the namepsace that gets generated for them.

> 
> | Namepsace scoped
> 
> 2. how would you implement setting limits to the amount of workspaces a user can create? since all the workspace objects exist in the same namespace, ResourceQuota can't be used (they are tied to a namespace, not a user).

You can have a two prone approach Validating Web Hook that limits the number of workspaces an owner can have. But also the controller reconciler can check and error out. This provide additional protection. Lets say the validating web hook crashes for whatever reason, well the reconciler will catch it. Lets say the reconciler crashes, well it won't be able to create the corresponding namespace, which is what we are trying to prevent in the first place. Kyverno can simplify the validating webhook part, but won't create that addition protection that the reconciler provides

> 
> | Is it introducing just a controller or also a REST service? | Both
> 
> 3. Today in upstream konflux the user can do everything he/she need using `kubectl`. how will it work with the workspaces rest service, can I create workspaces using kubectl?

Yes I would not have it other way, the REST API is fully K8s compatible and user will be able to interact with like they do with any other resource via kubectl or via GitOps if they want

Reply:
| Yes I would not have it other way, the REST API is fully K8s compatible and user will be able to interact with like they do with any other resource via kubectl or via GitOps if they want

it would mean the user will have to connect to the workspace rest service when he want to work with the workspace api and then switch to the cluster's api when he want to work with other resources?

> You can have a two prone approach Validating Web Hook that limits the number of workspaces an owner can have. But also the controller reconciler can check and error out. This provide additional protection. Lets say the validating web hook crashes for whatever reason, well the reconciler will catch it. Lets say the reconciler crashes, well it won't be able to create the corresponding namespace, which is what we are trying to prevent in the first place. Kyverno can simplify the validating webhook part, but won't create that addition protection that the reconciler provides

1. Is there any reason a validating webhook, such the one described above, can't be created for namespace objects?

2. If the validating webhook crashes it will block the creation of the resources it admits, since the k8s api will try to forward the request to a server that won't respond.

3. built in kubernets quotas are implemented using admission webhooks (https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#resourcequota), so I think it would be safe enough to relay solely on admission webhook for limiting the amount of namespaces a user can create.